### PR TITLE
[mkldnn] optimize for mkldnn batchnorm backward

### PR DESCRIPTION
**Pull request by rongzha1**
## Description ##
optimized for mkldnn bwd bn primitive.

(time in mkldnn verbose)/OP_profile enhanced from 83% to 96%, image/s from 65.2 to 66.5 in resnet50

@TaoLv @ElaineBao @PatricZhao 

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

**Comment by rongzha1** [comment]
@mxnet-bot run ci [all]

File: src/operator/nn/mkldnn/mkldnn_batch_norm-inl.h [hunk 1]
```diff
@@ -332,17 +332,20 @@ void MKLDNNBatchNormBackward(const nnvm::NodeAttrs &attrs, const OpContext &ctx,
...
+    size_t copy_size = sizeof(DType) * channels_;
+    if (!param.fix_gamma) {
+      memcpy(weight_buf, weight_ptr, copy_size);
+      memcpy(&weight_buf[channels_], bias_ptr, copy_size);
+    } else {
+      for (int i = 0; i < channels_; i++) {
         weight_buf[i] = static_cast<DType>(1.0f);
+      }
+      for (int i = 0; i < channels_; i++) {
+        weight_buf[channels_ + i] = (beta.data().dptr<DType>())[i];  // bias
```
Comment (ElaineBao) [inline src/operator/nn/mkldnn/mkldnn_batch_norm-inl.h]: Using `memcpy` as above

File: src/operator/nn/mkldnn/mkldnn_batch_norm-inl.h [hunk 2]
```diff
@@ -381,15 +384,19 @@ void MKLDNNBatchNormBackward(const nnvm::NodeAttrs &attrs, const OpContext &ctx,
...
-      (in_grad[2].data().dptr<DType>())[i] = gw_buf[i + channels_];
+    if (!param.fix_gamma) {
+      memcpy(w_grad_1, gw_buf, copy_size);
+      memcpy(w_grad_2, &gw_buf[channels_], copy_size);
+    } else {
+      for (int i = 0; i < channels_; i++) {
+        (in_grad[1].data().dptr<DType>())[i] = 0.0f;
+      }
+      for (int i = 0; i < channels_; i++) {
+        (in_grad[2].data().dptr<DType>())[i] = gw_buf[i + channels_];
```
Comment (ElaineBao) [inline src/operator/nn/mkldnn/mkldnn_batch_norm-inl.h]: Using `memcpy` as above.

File: src/operator/nn/mkldnn/mkldnn_batch_norm-inl.h (hunk 1 above)
Comment (rongzha1) [reply src/operator/nn/mkldnn/mkldnn_batch_norm-inl.h]: OK

File: src/operator/nn/mkldnn/mkldnn_batch_norm-inl.h (hunk 2 above)
Comment (rongzha1) [reply src/operator/nn/mkldnn/mkldnn_batch_norm-inl.h]: OK

**Comment by rongzha1** [comment]
@mxnet-bot run ci [all]

**Comment by rongzha1** [comment]
@mxnet-bot run ci windows-gpu

File: src/operator/nn/mkldnn/mkldnn_batch_norm-inl.h [hunk 3]
```diff
@@ -180,9 +180,10 @@ void MKLDNNBatchNormForward(const nnvm::NodeAttrs &attrs, const OpContext &ctx,
     CHECK(weight_mem.get_desc().get_size() == channels_ * sizeof(float) * 2);
     float* weight_ptr = gamma.data().dptr<float>();
     float* bias_ptr = beta.data().dptr<float>();
+    size_t copy_size = sizeof(weight_buf[0]) * channels_;
```
Comment (TaoLv) [inline src/operator/nn/mkldnn/mkldnn_batch_norm-inl.h]: const?

File: src/operator/nn/mkldnn/mkldnn_batch_norm-inl.h [hunk 4]
```diff
@@ -332,17 +333,18 @@ void MKLDNNBatchNormBackward(const nnvm::NodeAttrs &attrs, const OpContext &ctx,
     const NDArray &beta     = in_data[batchnorm::kBeta];
     DType *weight_buf = reinterpret_cast<DType *>(bwd.GetWeight().get_data_handle());
     nnvm::dim_t channels_ = data.shape()[1];
-    for (int i = 0; i < channels_; i++) {
-      if (!param.fix_gamma)
-        weight_buf[i] = (gamma.data().dptr<DType>())[i];   // weight
-      else
+    DType *weight_ptr = reinterpret_cast<DType *>(gamma.data().dptr<DType>());
+    DType* bias_ptr = reinterpret_cast<DType *>(beta.data().dptr<DType>());
```
Comment (TaoLv) [inline src/operator/nn/mkldnn/mkldnn_batch_norm-inl.h]: It seems `reinterpret_cast<DType*>` is not needed.

File: src/operator/nn/mkldnn/mkldnn_batch_norm-inl.h [hunk 5]
```diff
@@ -381,15 +383,17 @@ void MKLDNNBatchNormBackward(const nnvm::NodeAttrs &attrs, const OpContext &ctx,
...
     // copy data from gradw_mem to in_grad[1] and in_grad[2]
     DType *gw_buf = reinterpret_cast<DType *>(bwd.GetGradw().get_data_handle());
-    for (int i = 0; i < channels_; i++) {
-      if (!param.fix_gamma)
-        (in_grad[1].data().dptr<DType>())[i] = gw_buf[i];
-      else
-        (in_grad[1].data().dptr<DType>())[i] = 0.0f;
-    }
+    DType *w_grad_1 = reinterpret_cast<DType *>(in_grad[1].data().dptr<DType>());
+    DType *w_grad_2 = reinterpret_cast<DType *>(in_grad[2].data().dptr<DType>());
```
Comment (TaoLv) [inline src/operator/nn/mkldnn/mkldnn_batch_norm-inl.h]: Same here.

**Comment by rongzha1** [comment]
@mxnet-bot run ci [all]

File: src/operator/nn/mkldnn/mkldnn_batch_norm-inl.h (hunk 4 above)
Comment (rongzha1) [reply src/operator/nn/mkldnn/mkldnn_batch_norm-inl.h]: exactly. Done

File: src/operator/nn/mkldnn/mkldnn_batch_norm-inl.h (hunk 3 above)
Comment (rongzha1) [reply src/operator/nn/mkldnn/mkldnn_batch_norm-inl.h]: OK thanks

**Comment by rongzha1** [comment]
@mxnet-bot run ci [windows-gpu, unix-gpu]

**Comment by rongzha1** [comment]
@mxnet-bot run ci [windows-gpu, unix-gpu]

**Comment by rongzha1** [comment]
@mxnet-bot run ci [all]

**Comment by rongzha1** [comment]
Hi @pengzhao-intel @TaoLv @ChaiBapchya CI passed. please help review this PR again.

**Comment by pengzhao-intel** [comment]
Please file a new PR to 1.7 branch :)

**Pull request merged**